### DISCUSSION TITLE
HTTPonly cookies not being available after a request is made until page refresh - breaks `router.push`

### DISCUSSION THREAD

**Discussion by hawkyre** [vercel/next.js#38676] [Help]
I've got an external Elixir API which handles all the server stuff, including auth. When I make the signup request to it, an HTTPonly cookie is set with the user's session. However, I then perform a redirection to another specific page in my site, which only happens after signing up. However, that redirection is not performed properly since that route checks for the session and it does not find it for some reason, which causes it to redirect back to the sign in page. As soon as the user switches tabs or refreshes the page, however, the redirections happen as usual. In this case, the browser checks for a session in the signin and, if it finds it, it redirects the user to the dashboard.

Here's some of my code to showcase what's going on:

- Code that is executed when finishing the sign up
```ts
const onFinishSignup = async () => {
    const res = await APIClient.auth.signUp({...}); // Uses axios

    if (res.data?.ok) {
      router.push('/enroll?firstTime=true');
      router.reload(); // I tried putting this too but it does nothing
    } else {
      errorToast(
        'There was a problem when creating your account.'
      );
    }
  };
```

- Code that is executed when signing in
```ts
const submit: FormEventHandler<HTMLButtonElement | HTMLFormElement> = async (
    e
  ) => {
    setLoading(true);
    e.preventDefault();

    const result = await APIClient.auth.signIn(formData);

    if (result.error) {
      setError(result.error.title);
    } else {
      if (result.data?.enrolled) {
        router.push('/dashboard');
      } else {
        router.push('/enroll?firstTime=true');
      }

      router.reload();
    }

    setLoading(false);
    return false;
  };
```

*The router.reload calls don't affect the end result*

All the router.push calls here do not take the HTTPonly cookie into account and thus the redirects are performed as if the user had not been logged in.

Can anyone shine some light on this? Thanks!

**Comment by hawkyre** [comment]
Note: I want to make it clear that the cookie does work and I have put the `credentials: true` with axios, it just does not happen right after the request is made.

**Reply by mlima2412** [reply]
Hello! Have you solved this issue? I have the exactly problem right now. If you solved, can you help me please.

**Comment by icyJoseph** [comment]
Ah this issue has so many shapes and forms. Let's see if we can debug this one.

First of all, the cookie that the API sets, does it get this flag?

<img width="469" alt="Screenshot 2022-07-15 at 15 56 45" src="https://user-images.githubusercontent.com/21013447/179238207-436895db-5ffa-44b8-a6d3-a977acbc3721.png">

**Reply by hawkyre** [reply]
I can't see anything that resembles that flag, I've looked through all the relevant requests made.

**Comment by darushHamidi** [comment]
I have the same problem, 

After Login, I navigate the user to http://localhost:3000 (main page) then On the main page there is a button that you can hit it to navigate to the dashboard, Here It does not let you 
navigate, first I have to reload the page then I can navigate. This problem happens in production. (there is no problem in development)


NOTE:
Please change the URL by hitting the button.(do not do that by hand because you can not see the problem)

This is a small repo: https://github.com/darushHamidi/auth-with-middleware

Thanks in advance

**Reply by nilsfs7** [reply]
same issue for me. if you found any solution please let me know

**Reply by tomjamesallen** [reply]
related?
https://github.com/vercel/next.js/discussions/38650#discussioncomment-3148772
